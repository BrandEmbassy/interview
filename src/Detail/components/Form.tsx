import * as React from 'react';
import { CompState, editContact, changeVal, saveContact, deleteContact, InputType } from '../state';

const handleInputChange = (type: InputType) => (
  e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
) => {
  changeVal({ type, val: e.target.value });
};

const Form = (id: string, replace: (path: string) => void) => ({
  editMode,
  values,
}: CompState): JSX.Element => (
  <div className="detail">
    <div className="item">
      <div className="item__header">
        <div className="profile-pic" />
        <input
          className="name"
          type="text"
          name="name"
          value={values.name}
          placeholder="Full Name"
          disabled={!editMode}
          onChange={handleInputChange('name')}
        />
      </div>
      <div className="item__content">
        <div className="input-wrap">
          <label htmlFor="bio">Bio</label>
          <textarea
            name="bio"
            className="bio"
            placeholder="Decsription"
            value={values.bio}
            disabled={!editMode}
            onChange={handleInputChange('bio')}
          />
        </div>
        <div className="input-wrap">
          <label htmlFor="tel">Phone</label>
          <input
            type="text"
            name="tel"
            className="tel"
            value={values.tel}
            placeholder="+XXX XXX XXX XXX"
            disabled={!editMode}
            onChange={handleInputChange('tel')}
          />
        </div>
        <div className="input-wrap">
          <label htmlFor="email">E-mail</label>
          <input
            type="text"
            className="email"
            value={values.email}
            placeholder="E-mail"
            disabled={!editMode}
            onChange={handleInputChange('email')}
          />
        </div>
      </div>
      <div className="item__footer">
        {editMode ? (
          <div className="button button--positive" onClick={() => saveContact(Number(id))}>
            Save
          </div>
        ) : (
          <div className="button" onClick={() => editContact(Number(id))}>
            Edit
          </div>
        )}

        <div
          className="button button--negative"
          onClick={() => deleteContact({ id: Number(id), replace })}
        >
          Delete
        </div>
      </div>
    </div>
  </div>
);

export default Form;
