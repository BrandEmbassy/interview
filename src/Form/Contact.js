import util from '../util'

export default [
  {
    form: 'contact',
    fields: ["fullName", "bio", "tel", "email"],
    validate: (values) => {
      const errors = {}

      if (!values.fullName) {
        errors.fullName = 'Required'
      }

      if (!util.isPhone(values.tel)) {
        errors.tel = 'Wrong format'
      }

      if (!util.isEmail(values.email)) {
        errors.email = 'Wrong format'
      }

      return errors
    }
  },
  function({ contacts, selectedId }) {
    return {
      initialValues: contacts.find((contact) => contact.id === selectedId)
    }
  }
]
