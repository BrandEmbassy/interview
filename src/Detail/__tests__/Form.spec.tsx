import Form from '../components/Form';
import render from '../../test/render';

[
  {
    editMode: false,
    values: { name: '', bio: '', tel: '', email: '' },
  },
  {
    editMode: true,
    values: { name: '', bio: '', tel: '', email: '' },
  },
  {
    editMode: true,
    values: { name: 'fdsf', bio: 'fds', tel: '432', email: 'fds@fdsf.sdf' },
  },
].forEach(props => {
  it(`should render Detail - ${JSON.stringify(props, null, 2)}`, () => {
    const Comp = Form('1', () => {})(props);
    expect(render(Comp)).toMatchSnapshot();
  });
});
