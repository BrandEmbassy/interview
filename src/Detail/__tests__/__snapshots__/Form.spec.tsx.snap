// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Detail - {
  "editMode": false,
  "values": {
    "name": "",
    "bio": "",
    "tel": "",
    "email": ""
  }
} 1`] = `
<div
  className="detail"
>
  <div
    className="item"
  >
    <div
      className="item__header"
    >
      <div
        className="profile-pic"
      />
      <input
        className="name"
        disabled={true}
        name="name"
        onChange={[Function]}
        placeholder="Full Name"
        type="text"
        value=""
      />
    </div>
    <div
      className="item__content"
    >
      <div
        className="input-wrap"
      >
        <label
          htmlFor="bio"
        >
          Bio
        </label>
        <textarea
          className="bio"
          disabled={true}
          name="bio"
          onChange={[Function]}
          placeholder="Decsription"
          value=""
        />
      </div>
      <div
        className="input-wrap"
      >
        <label
          htmlFor="tel"
        >
          Phone
        </label>
        <input
          className="tel"
          disabled={true}
          name="tel"
          onChange={[Function]}
          placeholder="+XXX XXX XXX XXX"
          type="text"
          value=""
        />
      </div>
      <div
        className="input-wrap"
      >
        <label
          htmlFor="email"
        >
          E-mail
        </label>
        <input
          className="email"
          disabled={true}
          onChange={[Function]}
          placeholder="E-mail"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="item__footer"
    >
      <div
        className="button"
        onClick={[Function]}
      >
        Edit
      </div>
      <div
        className="button button--negative"
        onClick={[Function]}
      >
        Delete
      </div>
    </div>
  </div>
</div>
`;

exports[`should render Detail - {
  "editMode": true,
  "values": {
    "name": "",
    "bio": "",
    "tel": "",
    "email": ""
  }
} 1`] = `
<div
  className="detail"
>
  <div
    className="item"
  >
    <div
      className="item__header"
    >
      <div
        className="profile-pic"
      />
      <input
        className="name"
        disabled={false}
        name="name"
        onChange={[Function]}
        placeholder="Full Name"
        type="text"
        value=""
      />
    </div>
    <div
      className="item__content"
    >
      <div
        className="input-wrap"
      >
        <label
          htmlFor="bio"
        >
          Bio
        </label>
        <textarea
          className="bio"
          disabled={false}
          name="bio"
          onChange={[Function]}
          placeholder="Decsription"
          value=""
        />
      </div>
      <div
        className="input-wrap"
      >
        <label
          htmlFor="tel"
        >
          Phone
        </label>
        <input
          className="tel"
          disabled={false}
          name="tel"
          onChange={[Function]}
          placeholder="+XXX XXX XXX XXX"
          type="text"
          value=""
        />
      </div>
      <div
        className="input-wrap"
      >
        <label
          htmlFor="email"
        >
          E-mail
        </label>
        <input
          className="email"
          disabled={false}
          onChange={[Function]}
          placeholder="E-mail"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="item__footer"
    >
      <div
        className="button button--positive"
        onClick={[Function]}
      >
        Save
      </div>
      <div
        className="button button--negative"
        onClick={[Function]}
      >
        Delete
      </div>
    </div>
  </div>
</div>
`;

exports[`should render Detail - {
  "editMode": true,
  "values": {
    "name": "fdsf",
    "bio": "fds",
    "tel": "432",
    "email": "fds@fdsf.sdf"
  }
} 1`] = `
<div
  className="detail"
>
  <div
    className="item"
  >
    <div
      className="item__header"
    >
      <div
        className="profile-pic"
      />
      <input
        className="name"
        disabled={false}
        name="name"
        onChange={[Function]}
        placeholder="Full Name"
        type="text"
        value="fdsf"
      />
    </div>
    <div
      className="item__content"
    >
      <div
        className="input-wrap"
      >
        <label
          htmlFor="bio"
        >
          Bio
        </label>
        <textarea
          className="bio"
          disabled={false}
          name="bio"
          onChange={[Function]}
          placeholder="Decsription"
          value="fds"
        />
      </div>
      <div
        className="input-wrap"
      >
        <label
          htmlFor="tel"
        >
          Phone
        </label>
        <input
          className="tel"
          disabled={false}
          name="tel"
          onChange={[Function]}
          placeholder="+XXX XXX XXX XXX"
          type="text"
          value="432"
        />
      </div>
      <div
        className="input-wrap"
      >
        <label
          htmlFor="email"
        >
          E-mail
        </label>
        <input
          className="email"
          disabled={false}
          onChange={[Function]}
          placeholder="E-mail"
          type="text"
          value="fds@fdsf.sdf"
        />
      </div>
    </div>
    <div
      className="item__footer"
    >
      <div
        className="button button--positive"
        onClick={[Function]}
      >
        Save
      </div>
      <div
        className="button button--negative"
        onClick={[Function]}
      >
        Delete
      </div>
    </div>
  </div>
</div>
`;
